acme/api/revoke.go

```diff
@@ -0,0 +1,357 @@
+// authorized. If the certificate retrieved from the database doesn't belong to the Account,
+// the identifiers in the certificate are extracted and compared against the (valid) Authorizations
+// that are stored for the ACME Account. If these sets match, the Account is considered authorized
+// to revoke the certificate. If this check fails, the client will receive an unauthorized error.
+func (h *Handler) isAccountAuthorized(ctx context.Context, dbCert *acme.Certificate, certToBeRevoked *x509.Certificate, account *acme.Account) *acme.Error {
+	if !account.IsValid() {
+		return wrapUnauthorizedError(certToBeRevoked, nil, fmt.Sprintf("account '%s' has status '%s'", account.ID, account.Status), nil)
+	}
+	certificateBelongsToAccount := dbCert.AccountID == account.ID
+	if certificateBelongsToAccount {
+		return nil // return early; skip relatively expensive database check
+	}
+	requiredIdentifiers := extractIdentifiers(certToBeRevoked)
+	if len(requiredIdentifiers) == 0 {
+		return wrapUnauthorizedError(certToBeRevoked, nil, "cannot authorize revocation without providing identifiers to authorize", nil)
+	}
+	authzs, err := h.db.GetAuthorizationsByAccountID(ctx, account.ID)
+	if err != nil {
+		return acme.WrapErrorISE(err, "error retrieving authorizations for Account %s", account.ID)
+	}
+	authorizedIdentifiers := map[string]acme.Identifier{}
+	for _, authz := range authzs {
+		// Only valid Authorizations are included
+		if authz.Status != acme.StatusValid {
+			continue
+		}
+		authorizedIdentifiers[identifierKey(authz.Identifier)] = authz.Identifier
+	}
+	if len(authorizedIdentifiers) == 0 {
+		unauthorizedIdentifiers := []acme.Identifier{}
+		for _, identifier := range requiredIdentifiers {
+			unauthorizedIdentifiers = append(unauthorizedIdentifiers, identifier)
+		}
+		return wrapUnauthorizedError(certToBeRevoked, unauthorizedIdentifiers, fmt.Sprintf("account '%s' does not have valid authorizations", account.ID), nil)
+	}
+	unauthorizedIdentifiers := []acme.Identifier{}
+	for key := range requiredIdentifiers {
+		_, ok := authorizedIdentifiers[key]
+		if !ok {
+			unauthorizedIdentifiers = append(unauthorizedIdentifiers, requiredIdentifiers[key])
+		}
+	}
+	if len(unauthorizedIdentifiers) != 0 {
+		return wrapUnauthorizedError(certToBeRevoked, unauthorizedIdentifiers, fmt.Sprintf("account '%s' does not have authorizations for all identifiers", account.ID), nil)
+	}
+
+	return nil
+}
+
+// identifierKey creates a unique key for an ACME identifier using
+// the following format: ip|127.0.0.1; dns|*.example.com
+func identifierKey(identifier acme.Identifier) string {
+	if identifier.Type == acme.IP {
+		return "ip|" + identifier.Value
+	}
+	if identifier.Type == acme.DNS {
+		return "dns|" + identifier.Value
+	}
+	return "unsupported|" + identifier.Value
+}
+
+// extractIdentifiers extracts ACME identifiers from an x509 certificate and
+// creates a map from them. The map ensures that double SANs are deduplicated.
```

Comment:
double -> duplicated

Reply:
Fixed in 004fc054d570908479915306d98ca07761ccd491